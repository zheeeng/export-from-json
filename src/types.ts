export type ExportType = 'txt' | 'json' | 'csv' | 'xls' | 'xml' | 'css' | 'html' | 'tsv'

export const exportTypes: { [ET in ExportType]: ET } = {
    txt : 'txt',
    css : 'css',
    html : 'html',
    json : 'json',
    csv : 'csv',
    xls : 'xls',
    xml : 'xml',
    tsv : 'tsv'
}
