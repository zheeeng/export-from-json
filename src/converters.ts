import { isArray, getEntries, normalizeXMLName, indent, stripHTML, assert, getKeys } from './utils.js'

export function _createFieldsMapper (fields?: string[] | Record<string, string>) {
  if (
    !fields
    || isArray(fields) && !fields.length
    || !isArray(fields) && !getKeys(fields).length
  ) return (item: Record<string, unknown> | Array<Record<string, unknown>>) => item

  const mapper = isArray(fields)
    ? fields.reduce(
        (map, key) => ({ ...map, [key]: key}),
        Object.create(null) as Record<string, string>,
      )
    : fields

  return (item: Record<string, unknown> | Array<Record<string, unknown>>) => {
    if (isArray(item)) {
      return item
        .map(
          i => getEntries<unknown>(i).reduce(
            (map, [key, value]) => {
              if (key in mapper) {
                map[mapper[key]] = value
              }

              return map
            },
            Object.create(null) as Record<string, unknown>,
          ),
        )
        .filter(
          i => getKeys(i).length,
        )
    }

    return getEntries<unknown>(item).reduce(
      (map, [key, value]) => {
        if (key in mapper) {
          map[mapper[key]] = value
        }

        return map
      },
      Object.create(null) as Record<string, unknown>,
    )
  }
}

export function _prepareData (data: object | string): Record<string, unknown> {
  const MESSAGE_VALID_JSON_FAIL = 'Invalid export data. Please provide a valid JSON'
  try {
    return (typeof data === 'string' ? JSON.parse(data) : data) as Record<string, unknown>
  } catch {
    throw new Error(MESSAGE_VALID_JSON_FAIL)
  }
}

// TODO:: execute toSchema implicit converting
export function _createJSONData (
  data: object,
  replacer: ((key: string, value: any) => any) | Array<number | string> | null = null,
  space: string | number,
) {
  const MESSAGE_VALID_JSON_FAIL = 'Invalid export data. Please provide valid JSON object'
  try {
    return JSON.stringify(data, replacer as any, space)
  } catch {
    throw new Error(MESSAGE_VALID_JSON_FAIL)
  }
}

export interface ITableMap {
  [key: string]: string[],
}

export function _createTableMap (data: any[]): ITableMap {
  return data.map(getEntries).reduce(
    (tMap, rowKVs, rowIndex) =>
      rowKVs.reduce(
        (map, [key, value]) => {
          const columnValues = map[key] || Array.from({ length: data.length }).map(_ => '')
          columnValues[rowIndex] =
            (typeof value !== 'string' ? JSON.stringify(value) : value) || ''
          map[key] = columnValues

          return map
        },
        tMap,
      ),
    Object.create(null) as Partial<ITableMap>,
  ) as ITableMap
}

export interface ITableEntries extends Array<{ fieldName: string, fieldValues: string[] }> {}

export function _createTableEntries (
  tableMap: ITableMap,
  beforeTableEncode: (entries: ITableEntries) => ITableEntries = i => i,
): ITableEntries {
  return beforeTableEncode(getEntries(tableMap).map(([fieldName, fieldValues]) => ({
    fieldName,
    fieldValues,
  })))
}

// Rule: Fields that contain commas must begin and end with double quotes.
// Addition Rule: Fields that contain double quotes must begin and end with double quotes.
// Rule: Fields that contain line breaks must begin and end with double quotes
//       (not all programs support values with line breaks).
// Rule: All other fields do not require double quotes.
// Rule: Double quotes within values are represented by two contiguous double quotes.
function encloser (value: string) {
  const enclosingCharacter = /,|"|\n/.test(value) ? '"' : ''
  const escaped = value.replace(/"/g, '""')

  return `${enclosingCharacter}${escaped}${enclosingCharacter}`
}

type CreateCSVDataOptions = {
  beforeTableEncode?: (entries: ITableEntries) => ITableEntries,
  delimiter?: ',' | ';',
}

const defaultCreateCSVDataOption: Required<CreateCSVDataOptions> = { beforeTableEncode: i => i, delimiter: ',' }

// Reference: https://techterms.com/definition/csv
export function createCSVData (
  data: any[],
  options: CreateCSVDataOptions = {}
) {
  const { beforeTableEncode, delimiter } = { ...defaultCreateCSVDataOption, ...options }

  if (!data.length) return ''

  const tableMap = _createTableMap(data)
  const tableEntries = _createTableEntries(tableMap, beforeTableEncode)

  // Rule: Columns (fields) are separated by commas.
  // Rule: Rows are separated by line breaks (newline characters).
  const head = tableEntries.map(({ fieldName }) => fieldName).join(delimiter) + '\r\n'
  const columns = tableEntries.map(({ fieldValues }) => fieldValues)
    .map(column => column.map(encloser))
  const rows = columns.reduce(
    (mergedColumn, column) => mergedColumn.map((value, rowIndex) => `${value}${delimiter}${column[rowIndex]}`),
  )

  return head + rows.join('\r\n')
}

export function _renderTableHTMLText (
  data: any[],
  beforeTableEncode: (entries: ITableEntries) => ITableEntries,
) {
  assert(data.length > 0)

  const tableMap = _createTableMap(data)
  const tableEntries = _createTableEntries(tableMap, beforeTableEncode)
  const head = tableEntries.map(({ fieldName }) => fieldName)
    .join('</b></th><th><b>')
  const columns = tableEntries.map(({ fieldValues }) => fieldValues)
    .map(column => column.map(value => `<td>${value}</td>`))
  const rows = columns.reduce(
    (mergedColumn, column) => mergedColumn
      .map((value, rowIndex) => `${value}${column[rowIndex]}`),
  )

  return `
    <table>
      <thead>
        <tr><th><b>${head}</b></th></tr>
      </thead>
      <tbody>
        <tr>${rows.join(`</tr>
        <tr>`)}</tr>
      </tbody>
    </table>
  `
}

type CreateXLSDataOptions = {
  beforeTableEncode?: (entries: ITableEntries) => ITableEntries,
}

const defaultCreateXLSDataOptions: Required<CreateXLSDataOptions> = { beforeTableEncode: i => i }

export function createXLSData (data: any[], options?: CreateXLSDataOptions) {
  const { beforeTableEncode } = { ...defaultCreateXLSDataOptions, ...options }

  if (!data.length) return ''

  const content =
`<html>
  <head>
    <meta charset="UTF-8" />
  </head >
  <body>
    ${_renderTableHTMLText(data, beforeTableEncode)}
  </body>
</html >
`

  return content
}

export function createXMLData (data: object) {
  const content =

`<?xml version="1.0" encoding="utf-8"?><!DOCTYPE base>
${_renderXML(data, 'base')}
`

  return content
}

function _renderXML (data: any, tagName: string, arrayElementTag = 'element', spaces = 0): string {
  const tag = normalizeXMLName(tagName)
  const indentSpaces = indent(spaces)

  if (data === null || data === undefined) {
    return `${indentSpaces}<${tag} />`
  }

  const content = isArray(data)
    ? data.map(item => _renderXML(item, arrayElementTag, arrayElementTag, spaces + 2)).join('\n')
    : typeof data === 'object'
      ? getEntries(data as Record<string, unknown>)
        .map(([key, value]) => _renderXML(value, key, arrayElementTag, spaces + 2)).join('\n')
      : indentSpaces + '  ' + stripHTML(String(data))

  const contentWithWrapper =

`${indentSpaces}<${tag}>
${content}
${indentSpaces}</${tag}>`

  return contentWithWrapper
}
