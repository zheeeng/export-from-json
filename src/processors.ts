import type { ExportType } from './types.js'

export function generateDataURI (content: string, type: ExportType, byBlob: boolean): string {
  switch (type) {
    case 'txt': {
      const blobType = 'text/plain;charset=utf-8'

      if (byBlob) return URL.createObjectURL(new Blob([content], { type: blobType }))

      return `data:,${blobType}` + encodeURIComponent(content)
    }
    case 'css': {
      const blobType = 'text/css;charset=utf-8'

      if (byBlob) return URL.createObjectURL(new Blob([content], { type: blobType }))

      return `data:,${blobType}` + encodeURIComponent(content)
    }
    case 'html': {
      const blobType = 'text/html;charset=utf-8'

      if (byBlob) return URL.createObjectURL(new Blob([content], { type: blobType }))

      return `data:,${blobType}` + encodeURIComponent(content)
    }
    case 'json': {
      const blobType = 'text/json;charset=utf-8'

      if (byBlob) return URL.createObjectURL(new Blob([content], { type: blobType }))

      return `data:,${blobType}` + encodeURIComponent(content)
    }
    case 'csv': {
      const blobType = 'text/csv;charset=utf-8'

      if (byBlob) return URL.createObjectURL(new Blob([content], { type: blobType }))

      return `data:,${blobType}` + encodeURIComponent(content)
    }
    case 'xls': {
      const blobType = 'text/application/vnd.ms-excel;charset=utf-8'

      if (byBlob) return URL.createObjectURL(new Blob([content], { type: blobType }))

      return `data:,${blobType}` + encodeURIComponent(content)
    }
    case 'xml': {
      const blobType = 'text/application/xml;charset=utf-8'

      if (byBlob) return URL.createObjectURL(new Blob([content], { type: blobType }))

      return `data:,${blobType}` + encodeURIComponent(content)
    }
    default : {
      return ''
    }
  }
}

export function downloadFile (content: string, type: ExportType, fileName: string = 'download', byBlob = true): void {
  const dataURI = generateDataURI(content, type, byBlob)

  const anchor = document.createElement('a')
  anchor.href = dataURI

  anchor.download = fileName
  anchor.setAttribute('style', 'visibility:hidden')

  document.body.appendChild(anchor)
  anchor.dispatchEvent(
    new MouseEvent('click', {
      bubbles: false,
      cancelable: false,
      view: window,
    }),
  )
  document.body.removeChild(anchor)
}
