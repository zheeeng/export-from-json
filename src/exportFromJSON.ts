import { assert, isArray, normalizeFileName } from './utils.js'
import { downloadFile } from './processors.js'
import { _prepareData, _createJSONData, createCSVData, createXLSData, createXMLData, _createFieldsMapper } from './converters.js'
import { exportTypes, ExportType } from './types.js'
export interface IOption<R = void> {
  data: object | string
  fileName?: string
  extension?: string
  fileNameFormatter?: (name: string) => string
  fields?: string[] | Record<string, string>
  exportType?: ExportType
  replacer?: ((key: string, value: any) => any) | Array<number | string> | null
  space?: string | number
  processor?: (content: string, type: ExportType, fileName: string) => R
  withBOM?: boolean
  beforeTableEncode?: (
    tableRow: Array<{ fieldName: string, fieldValues: string[] }>,
  ) => Array<{ fieldName: string, fieldValues: string[]}>
}

function exportFromJSON<R = void> ({
  data,
  fileName = 'download',
  extension,
  fileNameFormatter = name => name.replace(/\s+/, '_'),
  fields,
  exportType = 'txt',
  replacer = null,
  space = 4,
  processor = downloadFile as any,
  withBOM = false,
  beforeTableEncode = (i) => i,
}: IOption<R>): R {
  const MESSAGE_IS_ARRAY_FAIL = 'Invalid export data. Please provide an array of objects'
  const MESSAGE_UNKNOWN_EXPORT_TYPE = `Can't export unknown data type ${exportType}.`
  const MESSAGE_FIELD_INVALID = `Can't export string data to ${exportType}.`

  if (typeof data === 'string') {
    switch (exportType) {
      case 'txt':
      case 'css':
      case 'html': {
          return processor(data, exportType, normalizeFileName(fileName, extension ?? exportType, fileNameFormatter))
        }
      default:
        throw new Error(MESSAGE_FIELD_INVALID)
    }
  }

  const fieldsMapper = _createFieldsMapper(fields)

  const safeData = fieldsMapper(_prepareData(data))

  const JSONData = _createJSONData(safeData, replacer, space)

  switch (exportType) {
    case 'txt':
    case 'css':
    case 'html': {
      return processor(JSONData, exportType, normalizeFileName(fileName, extension ?? exportType, fileNameFormatter))
    }
    case 'json': {
      return processor(JSONData, exportType, normalizeFileName(fileName, extension ?? 'json', fileNameFormatter))
    }
    case 'csv': {
      assert(isArray(safeData), MESSAGE_IS_ARRAY_FAIL)
      const BOM = '\ufeff'
      const CSVData = createCSVData(safeData, beforeTableEncode)
      const content = withBOM ? BOM + CSVData : CSVData

      return processor(content, exportType, normalizeFileName(fileName, extension ?? 'csv', fileNameFormatter))
    }
    case 'xls': {
      assert(isArray(safeData), MESSAGE_IS_ARRAY_FAIL)
      const content = createXLSData(safeData, beforeTableEncode)

      return processor(content, exportType, normalizeFileName(fileName, extension ?? 'xls', fileNameFormatter))
    }
    case 'xml': {
      const content = createXMLData(safeData)

      return processor(content, exportType, normalizeFileName(fileName, extension ?? 'xml', fileNameFormatter))
    }
    default:
      throw new Error(MESSAGE_UNKNOWN_EXPORT_TYPE)
  }
}

exportFromJSON.types = exportTypes

exportFromJSON.processors = { downloadFile }

export default exportFromJSON
