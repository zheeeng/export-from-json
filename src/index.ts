import exportFromJSON from './exportFromJSON.js'
import type { ExportType } from './types.js'

export type { ExportType }
export default exportFromJSON
