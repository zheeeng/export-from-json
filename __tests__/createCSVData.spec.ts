import { testTableItems, testTableItems2, beforeTableEncode } from './shared/data'
import { createCSVData } from '../src/converters'

test('createCSVData works correctly', () => {
  expect(createCSVData(testTableItems)).toMatchSnapshot()
  expect(createCSVData(testTableItems2)).toMatchSnapshot()
})

test('createCSVData works correctly with specifying beforeTableEncode option', () => {
  expect(createCSVData(testTableItems, { beforeTableEncode })).toMatchSnapshot()
})
test('createCSVData works correctly with delimiter option', () => {
  expect(createCSVData(testTableItems, { delimiter: ';' })).toMatchSnapshot()
})

test('createCSVData works with empty array', () => {
  expect(createCSVData([])).toEqual('')
})
