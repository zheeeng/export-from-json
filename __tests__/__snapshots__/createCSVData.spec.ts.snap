// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createCSVData works correctly 1`] = `
"greeting,foo,bar,a,b,c,d,e,f
hello,,,,,,,,
,foo,bar,,,,,,
,,,100,\\"[1,2,3]\\",\\"{\\"\\"x\\"\\":200,\\"\\"y\\"\\":300}\\",D,null,"
`;

exports[`createCSVData works correctly 2`] = `
"greeting,foo,bar,a,b,c,d,e,f
\\"hel\\"\\"lo\\",,,,,,,,
,foo,\\"b\\"\\"a\\"\\"r\\",,,,,,
,,,100,\\"[1,2,3]\\",\\"{\\"\\"x\\"\\":200,\\"\\"y\\"\\":300}\\",D,null,"
`;

exports[`createCSVData works correctly with delimiter option 1`] = `
"greeting;foo;bar;a;b;c;d;e;f
hello;;;;;;;;
;foo;bar;;;;;;
;;;100;\\"[1,2,3]\\";\\"{\\"\\"x\\"\\":200,\\"\\"y\\"\\":300}\\";D;null;"
`;

exports[`createCSVData works correctly with specifying beforeTableEncode option 1`] = `
"greeting,foo,bar,a,b,c,d,e,f
hello,0,0,0,0,0,0,0,0
0,foo,bar,0,0,0,0,0,0
0,0,0,1000,\\"[1,2,3]\\",\\"{\\"\\"x\\"\\":200,\\"\\"y\\"\\":300}\\",D,null,0"
`;
